src/main/java/com/capstone/service/impl/MentorSnapshotServiceImpl.java

```diff
@@ -0,0 +1,324 @@
+            throw new MentorRouteMappingException("Smart update failed", e);
+        }
+    }
+
+    @Override
+    @CacheEvict(value = {"mentors", "mentors-with-specializations", "mentors-search", "mentor-single"}, allEntries = true)
+    public MentorSnapshot assignSpecializationsToMentor(ProduceMentorEvent event) {
+        log.info("Assigning specializations to mentor for mentorId: {}", event.getVersapathUserId());
+
+        try {
+            // Find existing mentor WITH route mappings
+            Optional<MentorSnapshot> existingMentor = mentorSnapshotRepository.findByMentorIdWithRouteMappings(event.getVersapathUserId());
+
+            if (existingMentor.isEmpty()) {
+                throw new MentorNotFoundException(event.getVersapathUserId());
+            }
+
+            MentorSnapshot mentor = existingMentor.get();
+            log.info("Found existing mentor for assignment: {} with {} existing specialization mappings",
+                    mentor.getMentorId(), mentor.getMentorRouteMappings().size());
+
+            // Use existing smart update logic to assign specializations
+            if (event.getSpecializations() != null && !event.getSpecializations().isEmpty()) {
+                smartUpdateMentorRouteMappings(mentor, event.getSpecializations());
+
+                // Save the updated mentor
+                MentorSnapshot updatedMentor = mentorSnapshotRepository.save(mentor);
+
+                log.info("Successfully assigned {} specializations to mentor {}, total mappings now: {}",
+                        event.getSpecializations().size(), updatedMentor.getMentorId(),
+                        updatedMentor.getMentorRouteMappings().size());
+
+                return updatedMentor;
+            } else {
+                throw new MentorRouteMappingException("No specialization mappings provided for assignment");
+            }
+
+        } catch (MentorNotFoundException e) {
+            log.error("Mentor not found for assignment with ID: {}", event.getVersapathUserId(), e);
+            throw e;
+        } catch (Exception e) {
+            log.error("Unexpected error assigning specializations to mentor with ID: {}", event.getVersapathUserId(), e);
+            throw new MentorProcessingException("Failed to assign specializations to mentor", e);
+        }
+    }
+
+    private List<TalentRouteSnapshot> verifySpecializationRoutes(List<UUID> specializationRouteIds) {
+        List<TalentRouteSnapshot> verifiedRoutes = new ArrayList<>();
+
+        for (UUID routeId : specializationRouteIds) {
+            // Use TalentRouteSnapshotService to verify route exists
+            Optional<TalentRouteSnapshot> route = talentRouteSnapshotService.findByTalentRouteId(routeId);
+            if (route.isEmpty()) {
+                throw new TalentRouteNotFoundException(routeId);
+            }
+            verifiedRoutes.add(route.get());
+        }
+
+        return verifiedRoutes;
+    }
```

Comment:
Since you're using sql database, there is no need to do iterate checks when you could simply do something like this

```Java
// SELECT t.id FROM TalentRouteSnapshot t WHERE t.talentRouteId IN :ids
List<UUID> findByTalentRouteIdIn(Collection<UUID> ids);
```